python/tank/context.py

```diff
@@ -737,10 +737,133 @@ def from_entity(tk, entity_type, entity_id):
+    """
+    # perform validation of the entity dictionary:
+    if not isinstance(entity_dictionary, dict):
+        raise TankError("Cannot create a context from an empty or invalid entity dictionary!")
+    if "type" not in entity_dictionary:
+        raise TankError("Cannot create a context without an entity type!")
+    if "id" not in entity_dictionary:
+        raise TankError("Cannot create a context without an entity id!")
+    
+    # prep our context data structure
+    context = {
+        "tk": tk,
+        "project": None,
+        "entity": None,
+        "step": None,
+        "user": None,
+        "task": None,
+        "additional_entities": []
+    }
+
+    entity_type = entity_dictionary["type"]
+    entity_id = entity_dictionary["id"]
+
+    # try to determine the various entities from the entity dictionary:
+    project = None
+    entity = None
+    step = None
+    task = None
+    fallback_to_ctx_from_entity = False
     if entity_type == "Project":
-        # no need to set entity to point at project in this case
-        # that only produces double entries.
-        context["entity"] = None
+        # find entities for a project context
+        project = entity_dictionary
+    elif entity_type == "Task":
+        # find entities for a task context
+        task = entity_dictionary
+        if "project" not in task or "entity" not in task or "step" not in task:
+            fallback_to_ctx_from_entity = True
+        else:
+            project = task["project"]
+            entity = task["entity"]
+            step = task["step"]
+    elif entity_type in ["PublishedFile", "TankPublishedFile"]:
+        # special case handling for published files:
+        if entity_dictionary.get("task"):
+            # construct a task context
+            return from_entity_dictionary(tk, entity_dictionary["task"])
+        elif entity_dictionary.get("entity"):
+            # construct an entity context
+            return from_entity_dictionary(tk, entity_dictionary["entity"])
+        elif entity_dictionary.get("project"):
+            # construct project context
+            return from_entity_dictionary(tk, entity_dictionary["project"])
+        else:
+            # fall back on from_entity:
+            fallback_to_ctx_from_entity = True
+    else:
+        # find entities for an entity context
+        entity = entity_dictionary
+        if "project" not in entity:
+            fallback_to_ctx_from_entity = True
+        else:
+            project = entity["project"]
+
+    if not fallback_to_ctx_from_entity:
+        # clean up entities and populate context structure:
+        def _build_clean_entity(ent):
+            """
+            Ensure entity has id, type and name fields:
+            """
+            if "id" not in ent or "type" not in ent:
+                return None
+            ent_name = _get_entity_name(ent)
+            if ent_name == None:
+                return None
```

Comment:
What does this mean? A comment would be cool! If you pass in `{type: Shot, id: 123}` and there is no name found, `None` is returned? Is this an indication that we don't have it cached? 

Maybe a better docstring would help explain the expected inputs and outputs for this method!